### ISSUE TITLE
tag type: clean type declaration

### ISSUE THREAD

**Issue by m-stein** [cbe-project/cbe#29]
Currently the tag type is a mod type that is used for both request tags assigned by the client of the CBE and primitive tags that stand for the different pending operations that primitives correspond to (IO, WB, Sync_SB, etc.). The programmer is responsible for not mixing the two meanings. As far as I understand, there should rather be two dedicated tag types. The request type should be based on an individual numeric range (currently 0 .. 2**32 - 1) and opaque for the CBE library. The primitive tag should be based on a literal enumeration and not allow for other values than the enumerated ones.

**Comment by m-stein** [comment]
Unfortunately, It seems that the assertion "request tag => opaque client identifier && primitive tag => operation identifier" doesn't hold. For example, the sub-program ``Pool.Request_For_Tag`` compares the tag it receives (that should contain a normal primitive operation identifier AFAICS) directly to the tags in the requests it stores (that should contain opaque client identifiers AFAICS).

@cnuke Do I miss something here?

**Comment by cnuke** [comment]
You are right that both types are mixed. The original tag of the request is needed whenever the CBE library interacts with the client (frontend) and somehow needs to be preserved. At some point the request tag was stored within the primitive tag by the splitter module. 

*Generally speaking, each module will return the primitive in `Peek_Completed_Primitive` that was submitted in `Submit_Primitive` and only sets its own tag on primitives that were generated by the module.*

The splitter primitive gets passed on to the virtual-block-device module that performs the VBA → PBA translation. Eventually the translation is done and depending on the operation of the request (`READ` or `WRITE`) different modules have to acknowledge the request at the frontend.

So, to cleanly seperate the types, you have to identify all locations where `Primitive.Tag` is abused (fortunately that are not that many). At the same time or rather as a first step I suggest to start by cleaning up `Request_Primitive_Type` as it also contains `Tag` field and re-uses the normal tags.

In addition to changing the tag types, it would be nice if you could also replace the usage of the tag as key for looking up the request. I think using the `Pool`s item index should suffice and would decouple it from a malicous client's tag.

**Comment by m-stein** [comment]
Fixed on my cbe/merge_to_staging branch. Commits with the same message prefix can be squashed (I preserved them in case individual aspects shall be removed again).

Summary:
* The new ``Request.Tag_Type`` is used only for the tag the CBE client assigned to a request. These values have no functional meaning for the CBE anymore as the client can chose them arbitrarily.
* The new ``Pool_Index_Type`` identifies a request at the request pool and has no invalid value.
* The new ``Pool_Index_Slot_Type`` is a pool index combined with a boolean whether the index is valid.
* In the new ``Primitive.Object_Type.Pool_Idx_Slot`` primitives carry the pool index of the request they correspond to when necessary. Note, that there are valid primitives that don't carry a pool index which is why the member must be a slot.
* The new ``Primitive.Tag_Type`` is only used for the coloring of primitives ("Write_Back", "Splitter", "Cache", ...). Note, that there is still an invalid value for this type as tags are stored outside of primitives. But the commits ensure that a valid primitive will now always contains a valid tag.

Furthermore, as a precondition for these changes, I re-worked the former "front/back end request prim" in the CBE:

* The new ``Library.Event_Type`` identifies an external event that the CBE waits for to happen (for instance that the wrapper code supplies required client block-data). The value names of this type are very specific about the event that is expected and don't misuse the Tag_Type values anymore.
* The new ``Library.Object.Wait_For_Front/Back_End`` members replace the former Library.Object.Front/Back_End_Req_Prim and contain a member of Event_Type each. I hope this makes the inner workings of the CBE a little bit more clear.
